workflow/scripts/compare_pathways.py

```diff
@@ -32,52 +33,71 @@ def prepare(df):
 prepared_diffexp_y = prepare(diffexp_y)
 combined = (
     prepared_diffexp_x.join(prepared_diffexp_y, on=["Name"], suffix="_y")
-    .with_columns(pl.min_horizontal("Combined FDR", "Combined FDR_y").alias("fdr_min"))
-    .filter(pl.col("fdr_min") <= 0.05)
+    .with_columns(
+        pl.col("Combined FDR").cast(pl.Float64),
+        pl.col("Combined FDR_y").cast(pl.Float64),
+        pl.col("total perturbation accumulation").cast(pl.Float64),
+        pl.col("total perturbation accumulation_y").cast(pl.Float64),
+    )
+    .with_columns(pl.min_horizontal("Combined FDR", "Combined FDR_y").alias("min fdr"))
+    .filter(pl.col("min fdr") <= 0.05)
     .rename(
         {
             "total perturbation accumulation": effect_x,
             "total perturbation accumulation_y": effect_y,
-            "fdr_min": "min fdr",
         }
     )
     .collect()
 )
 
-effects = combined.select(pl.col(effect_x, effect_y))
-min_value = effects.min().min_horizontal()[0]
-max_value = effects.max().max_horizontal()[0]
-combined = combined.with_columns(
-    abs(pl.col(effect_x) - pl.col(effect_y)).alias("difference")
+if not combined.is_empty():
+    combined = (
+        combined.with_columns(
+            abs(pl.col(effect_x) - pl.col(effect_y)).alias("difference")
+        )
+        .with_columns(
+            (-pl.col("min fdr").log(base=10) * pl.col("difference")).alias("pi_value")
+        )
+        .sort(pl.col("pi_value").abs(), descending=True)
+    )
+else:
+    combined = combined.with_columns(
+        [pl.lit(None).alias("difference"), pl.lit(None).alias("pi_value")]
+    )
+combined_pd = combined.select(
+    pl.col(
+        "Name",
+        "min fdr",
+        effect_x,
+        effect_y,
+        "difference",
+        "pathway id",
+        "pi_value",
+    )
 )
-combined_sorted = combined.sort("difference", descending=True)
-combined_pd = combined_sorted.select(
-    pl.col("Name", "min fdr", effect_x, effect_y, "difference", "pathway id")
-).to_pandas()
-combined_pd.to_csv(snakemake.output[0], sep="\t", index=False)
-
-# we cannot use vegafusion here because it makes the point selection impossible since
-# it prunes the required ext_gene column
-# alt.data_transformers.enable("vegafusion")
-alt.data_transformers.disable_max_rows()
+combined_pd.to_pandas().to_csv(snakemake.output[0], sep="\t", index=False)
 
+df = combined_pd.select(pl.col("min fdr", effect_x, effect_y)).to_pandas()
+min_value = min(df[effect_x].min(), df[effect_y].min())
+max_value = max(df[effect_x].max(), df[effect_y].max())
```

Comment:
We used pandas because it's difficult to plot polars dfs with altair

Reply:
> We used pandas because it's difficult to plot polars dfs with altair

Ah, OK. But maybe we can try to only switch to pandas at the last possible moment, then.

> I did not use min_horizontal and max_horizontal because I wanted to have the overall minimum and maximum and not all rowwise minima and maxima.

OK, so I might have misunderstood the code. How about something like this, does this give what you want:

```
min_value = min(
        combined_pd.select(
            [effect_x, effect_y]
        ).min()
    )
```

It should select the two relevant columns, return the minimum value in each column, and then take the minimum across those minimums.

Reply:
It took quite some time but now we are retrieving the min and max via polars. The actual retrievement of the min and max wasnt the problem but much rather getting it to work with altair again.

Reply:
Sorry for the hassle. But really great that you found a way to get `altair` to work with a `polars` dataframe!

Reply:
Altair in general seems to work very nicely with polars as it supports input via the DataFrame Interchange Protocol. The main hurdle in the end was (again) typing as polars cannot infer types itself with the laziness.